spec/expression.dd

```diff
@@ -1623,44 +1623,54 @@ $(GNAME AssertExpression):
-        about the code that it may employ to produce more efficient code.
+    $(P The first $(I AssignExpression) must evaluate to true. If it does not, an $(I Assert Failure)
+    has occurred and the program enters an $(I Invalid State).
     )
 
-    $(P Programs for which $(I AssignExpression) is false are invalid.
-        Subsequent to such a false result, the program is in an invalid,
-        non-recoverable state.
+    $(P If the first $(I AssignExpression) consists entirely of compile time constants,
+    and evaluates to false, it is a special case; it
+    signifies that it is unreachable code.
+    Compile Time Function Execution (CTFE) is not attempted.
     )
 
-    $(P As a debugging tool, the compiler may insert checks to verify that
-        the condition indeed holds by evaluating $(I AssignExpression) at
-        runtime.  If it evaluates to a non-null class reference, the class
-        invariant is run.  Otherwise, if it evaluates to a non-null pointer to
-        a struct, the struct invariant is run.  Otherwise, if the result is
-        false, an $(D AssertError) is thrown.  If the result is true, then no
-        exception is thrown. In this way, if a bug in the code causes the
-        assertion to fail, execution is aborted, prompting the programmer to
-        fix the problem.
+    $(P $(I AssertExpression) has different semantics if it is in a
+    $(DDLINK spec/unittest, Unit Tests, $(D unittest)) or
+    $(DDLINK spec/contracts, Contract Programming, $(D in) contract).
     )
 
-    $(P It is implementation defined whether the $(I AssignExpression) is
-        evaluated at run time or not. Programs that rely on side effects of $(I
-        AssignExpression) are invalid.
+    $(P The second $(I AssignExpression), if present, must be implicitly
+        convertible to type $(D const(char)[]).
     )
 
-    $(P The result type of an assert expression is $(D void).  Asserts are
-        a fundamental part of the $(DDLINK spec/contracts, Contract Programming,
-        Contract Programming) support in D.
+    $(P If the first $(I AssignExpression) is a reference to a class instance for
+    which a $(DDSUBLINK spec/class, invariants, Class Invariant) exists, the $(I Class Invariant) must hold.
     )
 
-    $(P The second $(I AssignExpression), if present, must be implicitly
-        convertible to type $(D const(char)[]).
-        It is evaluated if the
-        result is false, and the string result is appended to the
-        $(D AssertError)'s message.)
+    $(P If the first $(I AssignExpression) is a pointer to a struct instance for
+    which a $(I Struct Invariant) exists, the $(I Struct Invariant) must hold.
+    )
+
+    $(P The type of an $(I AssertExpression) is $(D void).
+    )
+
+    $(UNDEFINED_BEHAVIOR Once in an $(I Invalid State) the behavior of the continuing execution
+    of the program is undefined.)
+
+    $(IMPLEMENTATION_DEFINED Whether the first $(I AssertExpression) is evaluated
```

Comment:
You never close this ... any decent macro language would have made it _a lot_ easier to spot the error ;-)
Also your PR needs to be rebased against `upstream/master`, s.t. it includes the new macros.

Reply:
thanks!